File: src/StronglyConnectedComponents.scala
import scala.io.Source._

/**
 * @author kperikov
 */
object StronglyConnectedComponents {

  var used: Array[Boolean] = null
  var order: List[Int] = null
  var components: List[Int] = null

  def findOrder(n: Int, graph: Array[List[Int]]) = {
    order = Nil
    used = new Array[Boolean](n + 1)
    for (i <- 1 to n) {
      if (!used(i)) {
        dfs1(i, graph)
      }
    }
  }

  def dfs1(vertex: Int, g: Array[List[Int]]): Unit = {
    used(vertex) = true
    for (v <- g(vertex)) {
      if (!used(v)) {
        dfs1(v, g)
      }
    }
    order = vertex :: order
  }

  def dfs2(vertex: Int, g: Array[List[Int]]): Unit = {
    used(vertex) = true
    components = vertex :: components
    for (v <- g(vertex)) {
      if (!used(v)) {
        dfs2(v, g)
      }
    }
  }

  def findComponents(n: Int, g: Array[List[Int]]): List[Int] = {
    used = new Array[Boolean](n + 1)
    components = Nil
    var ans: List[Int] = Nil
    for (i <- 1 to n) {
      val vertex = order(n - i)
      if (!used(vertex)) {
        dfs2(vertex, g)
        ans = components.size :: ans
        components = Nil
      }
    }
    ans
  }

  def main(args: Array[String]): Unit = {
    val lines = fromFile("E:/file.txt").getLines
    val n = readInt()
    val arr = new Array[List[Int]](n + 1)
    for (i <- 1 to n) {
      arr(i) = Nil
    }
    for (line <- lines) {
      val edge = line.split(" ").map(_.toInt).toList
      arr(edge(0)) = edge(1) :: arr(edge(0))
    }
    findOrder(n, arr)
    findComponents(n, arr).sorted.foreach(println)
  }
}

File: src/main/java/StronglyConnectedComponents.java
import java.io.*;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.StringTokenizer;

/**
 * Java version of Scala code
 */
public class StronglyConnectedComponents implements Runnable {

    PrintWriter out;
    BufferedReader br;
    StringTokenizer st;
    ArrayList<ArrayList<Integer>> g;
    ArrayList<ArrayList<Integer>> gr;
    ArrayList<Integer> order;
    ArrayList<Integer> components;
    boolean[] used;

    public static void main(String[] args) throws IOException {
        new Thread(new StronglyConnectedComponents()).start();
    }

    public String next() throws IOException {
        while (st == null || !st.hasMoreTokens()) {
            st = new StringTokenizer(br.readLine());
        }
        return st.nextToken();
    }

    public int nextInt() throws IOException {
        return Integer.parseInt(next());
    }

    public void run() {
        try {
            br = new BufferedReader(new InputStreamReader(new FileInputStream("E:/file.txt")));
            out = new PrintWriter(new BufferedOutputStream(System.out));
            solve();
            out.close();
        } catch (IOException e) {
            throw new IllegalStateException(e);
        }
    }

    void findOrder(int n) {
        Arrays.fill(used, false);
        order = new ArrayList<Integer>();
        for (int i = 0; i < n; ++i) {
            if (!used[i]) {
                dfs1(i);
            }
        }
    }

    void dfs1(int vertex) {
        used[vertex] = true;
        for (int i = 0; i < g.get(vertex).size(); ++i) {
            int v = g.get(vertex).get(i);
            if (!used[v]) {
                dfs1(v);
            }
        }
        order.add(vertex);
    }

    void dfs2(int vertex) {
        used[vertex] = true;
        components.add(vertex);
        for (int i = 0; i < gr.get(vertex).size(); ++i) {
            int v = gr.get(vertex).get(i);
            if (!used[v]) {
                dfs2(v);
            }
        }
    }

    ArrayList<Integer> findComponents(int n) {
        Arrays.fill(used, false);
        components = new ArrayList<Integer>();
        ArrayList<Integer> ans = new ArrayList<Integer>();
        for (int i = 0; i < n; ++i) {
            int vertex = order.get(n - i - 1);
            if (!used[vertex]) {
                dfs2(vertex);
                ans.add(components.size());
                components.clear();
            }
        }
        return ans;
    }

    public void solve() throws IOException {
        int n = nextInt();
        int m = nextInt();
        g = new ArrayList<ArrayList<Integer>>(n);
        gr = new ArrayList<ArrayList<Integer>>(n);
        used = new boolean[n];
        for (int i = 0; i < n; ++i) {
            g.add(new ArrayList<Integer>());
            gr.add(new ArrayList<Integer>());
        }
        for (int i = 0; i < m; ++i) {
            int a = nextInt() - 1;
            int b = nextInt() - 1;
            g.get(a).add(b);
            gr.get(b).add(a);
        }
        long begin = System.currentTimeMillis();
        findOrder(n);
        System.out.println("Find order in " + (System.currentTimeMillis() - begin) + " msec");
        begin = System.currentTimeMillis();
        ArrayList<Integer> comp = findComponents(n);
        Collections.sort(comp, Collections.reverseOrder());
        System.out.println("Find components in " + (System.currentTimeMillis() - begin) + " msec");
        System.out.print("Ans = ");
        for (int i = 0; i < Math.min(5, comp.size()); ++i) {
            System.out.print(comp.get(i) + ",");
        }
        System.out.println();

    }
}
